### DISCUSSION TITLE
NextJS with i18n support testing

### DISCUSSION THREAD

**Discussion by rdagostino** [cypress-io/cypress#24708] [Questions and Help]
I'm working `Next 13`, `next-i18next` and `Cypress 11` and developing an application that responds to API requests and displays different components/content based on the locale defined in our environment variables that are set during our pipeline build.

The locale is set at application start
```js
module.exports = {
  i18n: {
    defaultLocale: 'en-US',
    locales: ['en-US', 'en-GB'],
    localeDetection: false,
    domains: [
      {
        domain: getDomain(),
        defaultLocale: process.env.DEFAULT_LOCALE,
        http: process.env.NODE_ENV === 'development',
      },
    ],
  },
};
```

Based on the locale set we define a configuration for feature flags that we enable/disable accordingly. The features are utilized in two ways:

1. Through `getStaticProps|getServerSideProps` in which we get the flag values and pass them through to their respective React components. We take this approach to avoid flashing in the application when accessing pages directly.
2. A custom hook which will access the flags when needed

Then when using Cypress, I want to test the differences in displays based on that locale. However, I'm unable to find a solution to do this.

I've reached out on the Cypress discord and was given some direction to look at [mocking props](https://glebbahmutov.com/blog/mock-network-from-server/)

A small reproducible repo for this issue -> https://github.com/rdagostino/cypress-i18n

Any ideas or suggestions around this is greatly appreciated.

**Comment by portyu9** [comment]
Hi @rdagostino 

In this reproduction, `DEFAULT_LOCALE` is part of the **Next.js server configuration**, not something Cypress can change dynamically after the application has already started.

Your `next-i18next.config.js` builds the Next.js i18n configuration from:

```js
defaultLocale: process.env.DEFAULT_LOCALE
```

and that configuration is imported into `next.config.js`.

So changing `Cypress.env()` during an E2E test will not reconfigure the already-running Next.js application.

For an E2E test, test the locale through the same mechanism the application actually uses:

- **domain routing**: start the application with the desired domain/default-locale configuration and run Cypress against that host
- **sub-path routing**: visit the locale-specific URL, for example `/en-GB/...`
- if deployments genuinely use different `DEFAULT_LOCALE` values, run separate app/Cypress jobs for those configurations

Also note that your config contains:

```js
localeDetection: false
```

so changing the browser's `Accept-Language` header alone is intentionally not going to cause Next.js to select or redirect to a different locale. With that setting disabled, Next.js takes the locale from the configured domain or locale path.

For Component Testing, mocking/injecting translated props can make sense. For E2E, I would keep the real Next.js routing/configuration in the test because that is exactly the behavior you're trying to verify.

Please mark this answer as accepted if it helped, thank you.
